### PR TITLE
Provide initialization of Active Storage

### LINKED ISSUE #30102: Add --skip-active-storage and do so automatically when --skip-active-record is used
You should be able to get Rails without Active Storage, so let's implement the standard opt-out. And let's invoke this out-out when the --skip-active-record flag is set as well. You can't use Active Storage without Active Record.

### PR DESCRIPTION
**Pull request by bogdanvlviv**
https://github.com/rails/rails/pull/30098#issuecomment-320523143

Closes #30102

### CODE REVIEW

**Comment by bogdanvlviv** [comment]
After adding Active Storage by default, new apps take up more space.
I got the error `Errno::ENOSPC: No space left on device - sendfile` on travis-ci(https://travis-ci.org/rails/rails/builds/261890209) and my virtual machine.
The reason that we don't delete a generated app after finishing a test case, I looked at size `/tmp` and saw that it took up more than 4G space after `rails/railties$ bundle exec rake test`. 
Looks like we must remove a generated app after a test's finishing.

This commit fixes it https://github.com/rails/rails/pull/30101/commits/02b64f4ea05d5c4402cce64889ecf16cdb26909f, but now there is another problem, it takes up more time for execution, does anyone have any idea?

File: railties/lib/rails/generators/app_base.rb
Comment (rafaelfranca) [inline railties/lib/rails/generators/app_base.rb]: instead of type check better to check if the class respond to the value

File: railties/lib/rails/generators/rails/plugin/templates/rails/application.rb
Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/plugin/templates/rails/application.rb]: Why this was removed?

File: railties/lib/rails/generators/app_base.rb
Comment (bogdanvlviv) [inline railties/lib/rails/generators/app_base.rb]: I Agree. Thanks.

File: railties/lib/rails/generators/rails/plugin/templates/rails/application.rb
Comment (bogdanvlviv) [inline railties/lib/rails/generators/rails/plugin/templates/rails/application.rb]: Oops. I missed.

**Comment by rafaelfranca** [comment]
Looking good so far. Could you rebase and fix the tests?

**Comment by willnet** [comment]
requiring activestorage.js looks to need to be removed on use`--skip-active-storage`

https://github.com/rails/rails/blob/615998c4cb3a36b00f4894ec3bf390eb968d8c8b/railties/lib/rails/generators/rails/app/templates/app/assets/javascripts/application.js.tt#L15

**Comment by bogdanvlviv** [comment]
@willnet Thanks. Now it covered by this PR.

File: railties/lib/rails/generators/rails/app/app_generator.rb
Comment (yhirano55) [inline railties/lib/rails/generators/rails/app/app_generator.rb]: It seems that it does not need to consider skip flag here. If it enabled skip flag for active storage, all files will be removed on `#delete_storage_files_skipping_active_storege`.

Comment (bogdanvlviv) [inline railties/lib/rails/generators/rails/app/app_generator.rb]: thanks i just removed `#delete_storage_files_skipping_active_storege` as it isn't needed

**Review by rafaelfranca — APPROVED** [review]
Can you squash the commit that add  the CHANGELOG in the commit that add that feature?

**Comment by bogdanvlviv** [comment]
@rafaelfranca Thank you for the review!
Currently, i have to fix failed tests in `railties/test/railties/engine_test.rb`
Also, i think we can merge it only after fixing the `The job exceeded the maximum time limit for jobs, and has been terminated.` to prevent failing next builds on CI.
I've mentioned it here https://github.com/rails/rails/pull/30101#issuecomment-320755439 and in Campfire.

**Comment by rafaelfranca** [comment]
Maybe something in your PR is adding more time to the build since I don't see it happening on master. Could you check which change is doing that?

**Comment by bogdanvlviv** [comment]
@rafaelfranca 
It definitely my changes add more time to build tests, since `rails new` runs `rails active_storage:install` by default.

I've got two problems since adding that:
The first related to memory, it is described in the commit of this PR [Fix error `Errno::ENOSPC: No space left on device - sendfile`](https://github.com/rails/rails/pull/30101/commits/4e47ffbeae966dc1d5fd990ca09c5e59975e0522),
The second related to time, [`Rails::Generators::Testing::Behaviour#run_generator` skips Active St…
…orage by default](https://github.com/rails/rails/pull/30101/commits/f1342befdbe2ecc806fac7fda1896cfc9b464b21), but I'm not sure that is the right.

**Comment by bogdanvlviv** [comment]
@rafaelfranca 

https://github.com/rails/rails/pull/30101#issuecomment-320755439
https://github.com/rails/rails/pull/30101#issuecomment-323168547

This PR is blocked by long-time execution of the tests.
I opened new PR(https://github.com/rails/rails/pull/30304) that can help resolve the long-time test execution problem.


I've tested it in 2 ways:
- This PR + [Split up the railties tests for CI #30304](https://github.com/rails/rails/pull/30304)

  The result of tests execution: https://travis-ci.org/bogdanvlviv/ci-rails/builds/266504728
    - Average time of the build `GEM=railties TEST_DIR=application/**`: 38 min
    - Average time of the build `GEM=railties TEST_DIR=generators/**`: 13 min
    - Average time of the build `GEM=railties TEST_DIRS=.,command/**,commands/**,configuration/**,engine/**,railties/**,test_unit/**`: 14 min

- This PR without the commit [`Rails::Generators::Testing::Behaviour#run_generator` skips Active Storage by default](https://github.com/rails/rails/pull/30101/commits/cfe87422ca2bfa80fce914e044949dab13d176b7) + [Split up the railties tests for CI #30304](https://github.com/rails/rails/pull/30304)

  The result of tests execution: https://travis-ci.org/bogdanvlviv/ci-rails/builds/266505354:
    - Average time of the build `GEM=railties TEST_DIR=application/**`: 43 min
    - Average time of the build `GEM=railties TEST_DIR=generators/**`: 28 min
    - Average time of the build `GEM=railties TEST_DIRS=.,command/**,commands/**,configuration/**,engine/**,railties/**,test_unit/**`: 13 min

**Comment by bogdanvlviv** [comment]
I will continue to work on this when I have more info about nixing `rails active_storage:install`. See
https://github.com/rails/rails/pull/30280#issuecomment-324357554

**Comment by bogdanvlviv** [comment]
@rafaelfranca This PR is ready for review.

Just want to notice one more time that `railties`' tests became slower with execution of  `rails active_storage:install` for new apps by default.

Recently CI was falling for this PR because of  Travis CI' time limit. After merging this #30520(Thanks @matthewd), tests got time to be executed under this PR.

I'm not sure should we wait for replacing of `active_storage:install` rake task, see https://github.com/rails/rails/pull/30280#issuecomment-324357554
/cc @dhh

**Comment by bogdanvlviv** [comment]
The [test](https://travis-ci.org/rails/rails/jobs/285679144#L1801) is failed because
execution of `rails active_storage:install` can't respect option `--quiet`.
(Related to https://github.com/rails/rails/pull/30700  /cc @y-yagi )

What do you think about copying ActiveStorage' migration to `railties/lib/rails/generators/rails/app/templates/db/migrate/`?

**Comment by bogdanvlviv** [comment]
@georgeclaghorn Do you have any replacement for `rails active_storage:install`? https://github.com/rails/rails/pull/30378#issuecomment-324334758

**Comment by georgeclaghorn** [comment]
@sgrif was looking into an alternative to the installation task, if I remember correctly.

**Comment by y-yagi** [comment]
@bogdanvlviv 
I think it's a good to add `capture` option support to `execute_command`. As described below.

```diff 
diff --git a/railties/lib/rails/generators/actions.rb b/railties/lib/rails/generators/actions.rb
index 9800e57..ea85c49 100644
--- a/railties/lib/rails/generators/actions.rb
+++ b/railties/lib/rails/generators/actions.rb
@@ -292,7 +292,8 @@ def execute_command(executor, command, options = {}) # :doc:
           log executor, command
           env  = options[:env] || ENV["RAILS_ENV"] || "development"
           sudo = options[:sudo] && !Gem.win_platform? ? "sudo " : ""
-          in_root { run("#{sudo}#{extify(executor)} #{command} RAILS_ENV=#{env}", verbose: false) }
+          capture = options[:capture] || false
+          in_root { run("#{sudo}#{extify(executor)} #{command} RAILS_ENV=#{env}", verbose: false, capture: capture) }
         end
 
         # Add an extension to the given name based on the platform.
diff --git a/railties/lib/rails/generators/app_base.rb b/railties/lib/rails/generators/app_base.rb
index b33fd35..778fb83 100644
--- a/railties/lib/rails/generators/app_base.rb
+++ b/railties/lib/rails/generators/app_base.rb
@@ -461,7 +461,7 @@ def generate_spring_binstubs
 
       def run_active_storage
         unless skip_active_storage?
-          rails_command "active_storage:install"
+          rails_command "active_storage:install", capture: options[:quiet]
         end
       end
```

File: railties/lib/rails/generators/app_base.rb
Comment (y-yagi) [inline railties/lib/rails/generators/app_base.rb]: I think that specifying `skip_active_record` directly is more simpler.

**Comment by y-yagi** [comment]
We also need `skip_active_storage?` check in template of Gemfile.
https://github.com/rails/rails/blob/68a2888ef08faf38c0643921a8259beef5fbe630/railties/lib/rails/generators/rails/app/templates/Gemfile#L25

File: railties/test/isolation/abstract_unit.rb
Comment (y-yagi) [inline railties/test/isolation/abstract_unit.rb]: Why was this added?

Comment (bogdanvlviv) [reply railties/test/isolation/abstract_unit.rb]: Intentionally, see description of commit [Fix error `Errno::ENOSPC: No space left on device - sendfile`](https://github.com/rails/rails/pull/30101/commits/d6aea9e087cbcf09f3a7fb487e25f873d76f24a5)

File: railties/lib/rails/generators/app_base.rb
Comment (bogdanvlviv) [reply railties/lib/rails/generators/app_base.rb]: Yeah it is simpler, but I think it is more correctly use `skip_active_storage?`

**Comment by bogdanvlviv** [comment]
The tests are green!

Thanks @y-yagi

**Comment by rafaelfranca** [comment]
Could you squash the commits?

**Comment by bogdanvlviv** [comment]
@rafaelfranca done

**Comment by rafaelfranca** [comment]
That are still too many commits related to the same change that can be squashed.

**Comment by bogdanvlviv** [comment]
I squashed one more time.

**Pull request merged**
